cms/taxonomy/management/commands/sync_topics.py

```diff
@@ -0,0 +1,211 @@
+import logging
+from typing import Any
+
+import requests
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
+from django.core.management import BaseCommand
+
+from cms.taxonomy.models import Topic
+
+logger = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    """Topic Sync management command."""
+
+    def handle(self, *args: Any, **options: Any) -> None:
+        logger.info("Fetching topics from API...")
+        fetched_topics = _fetch_all_topics()
+        logger.info("Fetched %d topics", len(fetched_topics))
+
+        logger.info("Syncing topics...")
+        _sync_with_fetched_topics(fetched_topics)
+
+        logger.info("Checking for removed topics...")
+        _check_for_removed_topics({topic["id"] for topic in fetched_topics})
+
+        logger.info("Finished syncing topics.")
+
+
+def _fetch_all_topics() -> list[dict[str, str]]:
+    """Collect a complete list of topics and their subtopics by doing a
+    depth/breadth-first search using a stack of URLs.
+    """
+    topics = []
+
+    if not settings.DP_TOPIC_API_URL:
+        raise ImproperlyConfigured('"DP_TOPIC_API_URL" must be set')
+
+    # Build a stack of topics URLs and parent IDs
+    request_stack: list[tuple[str, str | None]] = [(f"{settings.DP_TOPIC_API_URL}/topics", None)]
+
+    # Use the stack of subtopic URls to iterate through, fetching all the subtopics
+    while request_stack:
+        url, parent_id = request_stack.pop()
+        raw_topics = _request_topics(url)
+
+        # Extract just the fields we need
+        subtopics = [
+            {
+                k: v
+                for k, v in raw_topic.items()
+                if k
+                in {
+                    "title",
+                    "id",
+                    "description",
+                }
+            }
+            for raw_topic in raw_topics
+        ]
+
+        for topic in subtopics:
+            if parent_id:
+                topic["parent_id"] = parent_id
+
+        # Extend the topics list, to build a flat list of topics and subtopics which contain their own parent IDs
+        topics.extend(subtopics)
+
+        # Add any subtopics URLs from the topics we just fetched to the stack
+        request_stack.extend(_extract_subtopic_links(raw_topics))
+
+    return topics
+
+
+def _request_topics(url: str) -> list[dict[str, Any]]:
+    """Fetch topics from the API and return the items from the response."""
+    topics_response = requests.get(url, timeout=30)
+    topics_response.raise_for_status()
+    raw_topics: list[dict[str, Any]] = topics_response.json().get("items", [])
```

Comment:
Minor: add a retry to handle transient errors (or a follow-up card). Currently, one failure exits the entire script.

Reply:
Good point, I prefer using one of the available libraries to handle this, do we have a preferred choice?

Reply:
We have a couple of options here.

1. Wrap our request in `try-except` and retry a fixed number of times with an exponential backoff.

2. We can use `requests` and `urllib3`. This provides a lot of things such as retry requests for specific status codes, connection errors and timeouts. It also has backoff support.

3. We can use some libraries such as [tenacity](https://github.com/jd/tenacity) and [backoff](https://github.com/litl/backoff). Where we can utilise the decorators they provide.

Reply:
As discussed, let's keep it simple and go with 2.

Reply:
https://jira.ons.gov.uk/browse/CMS-372?jql=project%20%3D%20CMS%20AND%20status%20%3D%20%22To%20Do%22%20AND%20resolution%20%3D%20Unresolved%20AND%20reporter%20in%20(currentUser())%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC